core/api_files.go

```diff
@@ -190,3 +190,33 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
 
 	pbJSON(g, http.StatusOK, keys)
 }
+
+// lsThreadFileTargetKeys godoc
+// @Summary File data at hash
+// @Description Returns raw data for file
+// @Tags files
+// @Produce application/octet-stream
+// @Param file hash string true "file hash"
+// @Success 200 {object} pb.Keys "keys"
+// @Failure 400 {string} string "Bad Request"
+// @Router /keys/{target} [get]
+func (a *api) getFileData(g *gin.Context) {
+	hash := g.Param("hash")
+
+	var err error
+	reader, _, err := a.node.FileData(hash)
+	if err != nil {
+		g.String(http.StatusBadRequest, err.Error())
+	}
+
+	reader.Seek(0, 0)
+
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		g.String(http.StatusBadRequest, err.Error())
+	}
+
+	contentType := http.DetectContentType(data)
```

Comment:
I think  you can do this faster by first getting the file index (what we're going to call file meta data) object, which already has the mime type and content length of the underlying data. Then you can use `g.DataFromReader`, which means that you don't need to read the entire data into memory before serving it back.